lib/vnet/app_resolver.go

```diff
@@ -0,0 +1,178 @@
+// be implemented differently for Connect and `tsh vnet`.
+type AppProvider interface {
+	// ListProfiles lists the names of all profiles saved for the user.
+	ListProfiles() ([]string, error)
+
+	// GetCachedClient returns a [*client.ClusterClient] for the given profile and leaf cluster.
+	// [leafClusterName] may be empty when requesting a client for the root cluster. Returned clients are
+	// expected to be cached, as this may be called frequently.
+	GetCachedClient(ctx context.Context, profileName, leafClusterName string) (*client.ClusterClient, error)
+}
+
+// TCPAppResolver implements [TCPHandlerResolver] for Teleport TCP apps.
+type TCPAppResolver struct {
+	appProvider AppProvider
+	slog        *slog.Logger
+}
+
+// NewTCPAppResolver returns a new *TCPAppResolver which will resolve full-qualified domain names to
+// TCPHandlers that will proxy TCP connection to Teleport TCP apps.
+//
+// It uses [appProvider] to list and retrieve cluster clients which are expected to be cached to avoid
+// repeated/unnecessary dials to the cluster. These clients are then used to list TCP apps that should be
+// handled.
+//
+// [appProvider] is also used to get app certificates used to dial the apps.
+func NewTCPAppResolver(appProvider AppProvider) *TCPAppResolver {
+	return &TCPAppResolver{
+		appProvider: appProvider,
+		slog:        slog.With(teleport.ComponentKey, "VNet.AppResolver"),
+	}
+}
+
+// ResolveTCPHandler resolves a fully-qualified domain name (FQDN) to a TCPHandler for handling TCP
+// connections to a Teleport TCP app. It returns the TCPHandler if a matching app is found, else it will
+// return with match == false. Errors must only be returned for truly unexpected errors that should cause a
+// DNS request to fail.
+func (r *TCPAppResolver) ResolveTCPHandler(ctx context.Context, fqdn string) (handler TCPHandler, match bool, err error) {
+	profileNames, err := r.appProvider.ListProfiles()
+	if err != nil {
+		return nil, false, trace.Wrap(err, "listing profiles")
+	}
+	for _, profileName := range profileNames {
+		if fqdn == fullyQualify(profileName) {
+			// This is a query for the proxy address, which we'll never want to handle.
+			return nil, false, nil
+		}
+		if !isSubdomain(fqdn, profileName) {
+			// TODO(nklaassen): support leaf clusters and custom DNS zones.
+			continue
+		}
+
+		slog := r.slog.With("profile", profileName, "fqdn", fqdn)
+		rootClient, err := r.appProvider.GetCachedClient(ctx, profileName, "")
+		if err != nil {
+			// The user might be logged out from this one cluster (and retryWithRelogin isn't working). Don't
+			// return an error so that DNS resolution will be forwarded upstream instead of failing, to avoid
+			// breaking e.g. web app access (we don't know if this is a web or TCP app yet because we can't
+			// log in).
+			slog.InfoContext(ctx, "Failed to get teleport client.", "error", err)
+			continue
```

Comment:
once we start supporting custom DNS zones (https://github.com/gravitational/teleport/pull/41545) this is what we want, since multiple clusters could have the same custom zone